src/coreclr/src/jit/codegenxarch.cpp

```diff
@@ -4082,16 +4083,27 @@ void CodeGen::genCodeForShift(GenTree* tree)
         }
         else
         {
-            // First, move the operand to the destination register and
-            // later on perform the shift in-place.
-            // (LSRA will try to avoid this situation through preferencing.)
-            if (tree->GetRegNum() != operandReg)
+            int typeWidth    = genTypeSize(genActualType(targetType)) * 8;
+            int shiftByValue = (int)shiftBy->AsIntConCommon()->IconValue();
+            if (compiler->compExactlyDependsOn(InstructionSet_BMI2) && tree->OperIs(GT_ROL) && (shiftByValue > 0) &&
```

Comment:
You want `compOpportunisticallyDependsOn` instead.

`ExactlyDependsOn` implies that the checked instruction set missing will result in invalid behavior or potentially detrimental perf. While `OpportunisticallyDependsOn` just implies that we are checking for a more efficient pattern.

https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/vectors-and-intrinsics.md#mechanisms-in-the-jit-to-generate-correct-code-to-handle-varied-instruction-set-support has more details about the available checks and the requirements around code in S.P.Corelib vs other assemblies.